### ISSUE TITLE
`option.to` breaks the stream and exits the program

### ISSUE THREAD

**Issue by dnafication** [adaltas/node-csv#333]
**Describe the bug**

Using `csv-parse` package to parse a readable stream. Basically [this example](https://csv.js.org/parse/recipes/promises/) from your documentation. Introducing the `to` or `to_line` option seems to exit the stream without executing anything after `await finished(parser);` in line 27 below:

```js
import assert from 'assert'
import fs from 'fs'
import os from 'os'
import { parse } from 'csv-parse'
import { finished } from 'stream/promises'

// Prepare the dataset
await fs.promises.writeFile(
  `${os.tmpdir()}/input.csv`,
  ['a,b,c', '1,2,3', '3,4,5'].join('\n')
)
// Read and process the CSV file
const processFile = async () => {
  const records = []
  const parser = fs.createReadStream(`${os.tmpdir()}/input.csv`).pipe(
    parse({
      to: 2,   // <-- add this option
    })
  )
  parser.on('readable', function () {
    let record
    while ((record = parser.read()) !== null) {
      // Work with each record
      records.push(record)
    }
  })
  await finished(parser)
  // nothing gets executed after this
  console.log(records) // <-- this does not print anything
  return records
}
// Parse the CSV content
const records = await processFile()
console.log(records) // <-- this doesn't get executed

// Validate the records
assert.deepStrictEqual(records, [
  ['a', 'b', 'c'],
  ['1', '2', '3'],
])
```

**To Reproduce**

The example is provided in the description above.

**Additional context**

**Comment by foghina** [comment]
I'm having the same issue, here's my slightly shorter repro:

```
import * as fs from "node:fs";
import * as stream from "node:stream/promises";

import { parse } from "csv-parse";

const parser = fs.createReadStream("data.csv").pipe(
  parse({
    // to_line: 3
  })
);
parser.on("readable", async () => {
  let record;
  while ((record = parser.read()) !== null) {
    console.log(record);
  }
});
console.info("waiting for stream to finish...");
await stream.finished(parser);
console.info("this won't get logged if you uncomment the to_line option.");
```

**Comment by foghina** [comment]
FWIW, `parser.on('end', callback)` still gets invoked correctly, so that's one potential way around this issue for now, but it's still quite blocking.

**Comment by foghina** [comment]
Follow-up from my previous comment, one way around it is adding:

```js
parser.on('end', parser.destroy);
```

This will correctly resolve the promise and the code after `await` gets executed.

**Comment by wdavidw** [comment]
I am looking at it as well, trying to read the official doc and extract some sense.

Using the more robust strategy of try-and-fail, I get to the same conclusion, `destroy` triggers `stream.finished` and it must be called after `end` to avoid `ERR_STREAM_PREMATURE_CLOSE`. Not sure why, nor about the correct way to handle this behavior.

**Comment by wdavidw** [comment]
Instead of

```
parser.on('end', parser.destroy);
```

Try

```
parser.end()
parser.on('end', parser.destroy);
```

This seems to prevent the `ERR_STREAM_PREMATURE_CLOSE` error in my test

**Comment by wdavidw** [comment]
It is also associated with the `close` event which is not thrown unless `destroy` is called or unless there is an error.

**Comment by foghina** [comment]
Calling `parser.end()` before `parser.on()` seems to end the parser before any records get processed though? Since the `await stream.finished()` has to come after both those lines.

**Comment by foghina** [comment]
OK, `parser.on("end", parser.end)` seems to fix the premature error quite nicely.

**Comment by foghina** [comment]
Thanks for the fix! Just for posterity of before a new release is made, I realized that using `for await` also seemingly fixes the problem (unsure if any dangling streams are left around though). My code is now:

```js
const parser = fs.createReadStream("data.csv").pipe(
  parse({
    to_line: 2,
  })
);
for await (const record of parser) {
  console.log(record);
}
console.info("done.");
```

Which seems to work correctly without any errors/exceptions.

**Comment by wdavidw** [comment]
The all stream API is hard to grasp. I am not even sure whether the `close` event is expected to be called or not.

Anyway, new versions are published:

- csv-parse: 5.3.3 => 5.3.4
- csv: 6.2.5 => 6.2.6

**Comment by Yoshiji** [comment]
I am updating my npm dependencies, and this change in https://github.com/adaltas/node-csv/commit/ca3f55b7cf556b45377677428783608a2d9ebbb2 is breaking my CSV parsing done client-side in the browser. It seems that `parser.destroy` is `undefined`, leading to a type error: https://codesandbox.io/s/jolly-matan-5y0er4?file=/src/index.js

**Comment by wdavidw** [comment]
Can you go inside the source code of the parser and add a condition, for example

```js
if (this.destroy !== undefined) {
   this.on('end', this.destroy);
}
```

I would also need sth in ./demo to replicate the issue.

**Comment by wdavidw** [comment]
I am re-opening this issue. I did some testing after issue #410.

Issue #333 is fixed with `this.end(); this.destroy();`
Issue #410 is fixed with `this.end(); this.on('end', this.destroy);`

1. It happens that to fix #410 which uses async iterator, we need to partially break this issue which uses `stream.finished`.
2. By fixing #410, we now have two very similar tests with different outcome. One use a readable stream implemented with csv-generate, it works. One use a readable stream implemented with stream.Readable, it breaks.
3. The code that fixes issue #333 is no longer working with the latest Node.js version 21.4.0.

Here is the note I left inside the [source code](https://github.com/adaltas/node-csv/blob/master/packages/csv-parse/lib/index.js#L35-L45). I have the feeling that this could come from Node.js and be out of grasp. We welcome anyone who would stop by with the expertise to help us.

```js
this.end()
// Fix #333 and break #410
//   ko: api.stream.iterator.coffee
//   ko with v21.4.0, ok with node v20.5.1: api.stream.finished # aborted (with generate())
//   ko: api.stream.finished # aborted (with Readable)
// this.destroy()
// Fix #410 and partially break #333
//   ok: api.stream.iterator.coffee
//   ok: api.stream.finished # aborted (with generate())
//   broken: api.stream.finished # aborted (with Readable)
this.on('end', this.destroy);
```

I took the decision to fix issue #411, re-open this issue and release the patch version 5.5.3. Motivation for the patch SemVer is that this issue has not been fixed properly and was broken with the upcoming Node.js versions. I acknowledge this is a very controversial decision in term of SemVer.
